### PR TITLE
Handle the subject query string when encoding/decoding e-mails

### LINKED ISSUE #2149: Cannot save mailto: URLs with more than 88 characters
**Affected version(s)**

Contao **4.4.51**, Contao **4.9.5**

**Description**

If you create a new hyperlink content element and use the following as the **Link target**:

```
mailto:info@mailadresse.ch?subject=aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
```

the field will be blank after saving. However, if you remove one character from the subject, e.g.

```
mailto:info@mailadresse.ch?subject=aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
```

it works. As far as I tested it, it only depends on the length, not the contents of the query parameters.

Slack thread: https://contao.slack.com/archives/CK4J0KNDB/p1597226512483700

### PR DESCRIPTION
**Pull request by leofeyer**
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #2149
| Docs PR or issue | -

### CODE REVIEW

**Review by fritzmg — COMMENTED** [review]
What about `?subject=foo&body=foo` or just `?body=foo`?

**Comment by leofeyer** [comment]
Apparently, it can get way more complex than I thought. You can also specify multiple recipients or add CC and BCC recipients via query string:

https://blog.kulturbanause.de/2012/01/e-mail-links-mailto-mit-betreff-kopie-cc-blindkopie-bcc-und-nachrichteninhalt/

**Comment by fritzmg** [comment]
Interesting, didn't know that either. May be just split the string to the first occurrence of `?`?

**Comment by leofeyer** [comment]
It is not that easy, because `?` can be in the local part of the e-mail address, e.g. `"foo@bar?a"@example.org`. There are more unusual but valid examples in our validator test class:

https://github.com/contao/contao/blob/fce0586b51d0f8dc1866b83bf41b723b00b5e547/core-bundle/tests/Contao/ValidatorTest.php#L59-L62

**Comment by Toflar** [comment]
Some clients even supported attachments back in the day with which you could do funny things like `&attach=~/.ssh/id_rsa` and stuff. That's why it's not recommended to support all of them.

**Comment by leofeyer** [comment]
I have optimized the check in 26d2b1838761514ad82d7d699ec2bb27ea831f81. It now only checks the hostname for a query string, which cannot include a `?` character, so if one is present, we know it has to be a query string.

**Review by fritzmg — DISMISSED** [review]

File: core-bundle/src/Resources/contao/library/Contao/Idna.php [hunk 1]
```diff
@@ -101,14 +101,29 @@ public static function encodeEmail($strEmail)
...
 			return '';
 		}
 
-		return implode('@', $arrChunks) . '@' . $strHost;
+		$strQuery = null;
+
+		// Strip the query string (see #2149)
+		if (strpos($strHost, '?') !== false)
+		{
+			list($strHost, $strQuery) = explode('?', $strHost);
```
Comment (ausi) [inline core-bundle/src/Resources/contao/library/Contao/Idna.php]:
```suggestion
			list($strHost, $strQuery) = explode('?', $strHost, 2);
```

This way we would not strip occurrences of `?`s in the query itself.

File: core-bundle/src/Resources/contao/library/Contao/Idna.php [hunk 2]
```diff
@@ -131,14 +146,29 @@ public static function decodeEmail($strEmail)
...
+		{
+			return '';
+		}
+
+		$strQuery = null;
+
+		// Strip the query string (see #2149)
+		if (strpos($strHost, '?') !== false)
+		{
+			list($strHost, $strQuery) = explode('?', $strHost);
```
Comment (ausi) [inline core-bundle/src/Resources/contao/library/Contao/Idna.php]:
```suggestion
			list($strHost, $strQuery) = explode('?', $strHost, 2);
```

Comment (leofeyer) [reply core-bundle/src/Resources/contao/library/Contao/Idna.php]: Sorry, I should have thought of this. 🙄

**Review by ausi — APPROVED** [review]

**Pull request merged**